crates/typst/src/geom/gradient.rs

```diff
@@ -0,0 +1,1070 @@
+                    Quadrant::Fourth => angle + TAU,
+                };
+
+                let (sin, cos) = angle.sin_cos();
+
+                let length = sin.abs() + cos.abs();
+                if angle > FRAC_PI_2 && angle < 3.0 * FRAC_PI_2 {
+                    x = 1.0 - x;
+                }
+
+                if angle > PI {
+                    y = 1.0 - y;
+                }
+
+                (x as f64 * cos.abs() + y as f64 * sin.abs()) / length
+            }
+        };
+
+        self.sample(RatioOrAngle::Ratio(Ratio::new(t)))
+    }
+
+    pub fn anti_alias(&self) -> bool {
+        match self {
+            Self::Linear(linear) => linear.anti_alias,
+        }
+    }
+
+    /// Returns the relative placement of this gradient, handling
+    /// the special case of `Auto`.
+    pub fn unwrap_relative(&self, on_text: bool) -> Relative {
+        self.relative().unwrap_or_else(|| {
+            if on_text {
+                Relative::Parent
+            } else {
+                Relative::This
+            }
+        })
+    }
+}
+
+#[derive(Clone, Hash, PartialEq, Eq)]
+pub struct LinearGradient {
+    /// The color stops of this gradient
+    pub stops: Vec<(Color, Ratio)>,
+
+    /// The direction of this gradient
+    pub angle: Angle,
+
+    /// The color space in which to interpolate the gradient
+    pub space: ColorSpace,
+
+    /// The relative placement of the gradient
+    pub relative: Smart<Relative>,
+
+    /// Whether to anti-alias the gradient (used for sharp gradient)
+    pub anti_alias: bool,
+}
+
+impl Debug for LinearGradient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("gradient.linear(")?;
+
+        let angle = self.angle.to_rad().rem_euclid(TAU);
+        if angle.abs() < EPSILON {
+            // Default value, do nothing
+        } else if (angle - FRAC_PI_2).abs() < EPSILON {
+            f.write_str("dir: rtl, ")?;
+        } else if (angle - PI).abs() < EPSILON {
+            f.write_str("dir: ttb, ")?;
+        } else if (angle - 3.0 * FRAC_PI_2).abs() < EPSILON {
+            f.write_str("dir: btt, ")?;
+        } else {
+            write!(f, "angle: {:?}, ", self.angle)?;
+        }
+
+        if self.space != ColorSpace::Oklab {
+            write!(f, "space: {:?}, ", self.space.into_value())?;
+        }
+
+        if self.relative.is_custom() {
+            write!(f, "relative: {:?}, ", self.relative.into_value())?;
+        }
+
+        for (i, (color, offset)) in self.stops.iter().enumerate() {
+            write!(f, "({color:?}, {offset:?})")?;
+
+            if i != self.stops.len() - 1 {
+                f.write_str(", ")?;
+            }
+        }
+
+        f.write_char(')')
+    }
+}
+
+/// What is the gradient relative to:
+/// - Itself (its own bounding box)
+/// - Its parent (the parent's bounding box)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Relative {
+    This,
+    Parent,
+}
```

Comment:
Should produce more useful documentation hints when coding

```suggestion
/// What is the gradient relative to.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Relative {
    /// The gradient is relative to itself (its own bounding box).
    This,
    
    /// The gradient is relative to its parent (the parent's bounding box).
    Parent,
}
```